File: crates/flowrt-codegen/tests/golden/fault_injection_restart_cpp/expected/cpp/tests/message_abi.cpp
// FlowRT 管理产物。不要手工修改。

#include <array>
#include <cassert>
#include <cstddef>
#include <cstdint>
#include <cstring>
#include <filesystem>
#include <fstream>
#include <stdexcept>
#include <string>
#include <string_view>
#include <type_traits>

#include "flowrt_app/messages.hpp"

namespace {

template <typename T>
std::array<std::uint8_t, sizeof(T)> bytes_of(const T& value) {
    std::array<std::uint8_t, sizeof(T)> bytes{};
    std::memcpy(bytes.data(), &value, bytes.size());
    return bytes;
}

template <typename T>
void assert_default_bytes_zero() {
    T value{};
    std::array<std::uint8_t, sizeof(T)> expected{};
    assert(bytes_of(value) == expected);
}

template <typename T>
void assert_byte_roundtrip(const T& value) {
    const auto bytes = bytes_of(value);
    T roundtrip{};
    std::memset(&roundtrip, 0, sizeof(roundtrip));
    std::memcpy(&roundtrip, bytes.data(), bytes.size());
    assert(std::memcmp(&roundtrip, &value, sizeof(T)) == 0);
}

template <typename T, std::size_t N>
void assert_sample_bytes(const T& value, const std::array<std::uint8_t, N>& expected) {
    static_assert(sizeof(T) == N);
    assert(bytes_of(value) == expected);
}

template <std::size_t N>
void write_fixture(std::string_view name, const std::array<std::uint8_t, N>& bytes) {
#ifdef FLOWRT_ABI_FIXTURE_DIR
    std::filesystem::create_directories(FLOWRT_ABI_FIXTURE_DIR);
    auto path = std::filesystem::path(FLOWRT_ABI_FIXTURE_DIR) / std::string(name);
    std::ofstream output(path, std::ios::binary);
    if (!output) {
        throw std::runtime_error("failed to open ABI fixture output");
    }
    output.write(reinterpret_cast<const char*>(bytes.data()), static_cast<std::streamsize>(bytes.size()));
    if (!output) {
        throw std::runtime_error("failed to write ABI fixture output");
    }
#else
    (void)name;
    (void)bytes;
#endif
}

constexpr std::array<std::uint8_t, 4> EXPECTED_SAMPLE_BYTES{{2, 0, 0, 0}};

flowrt_app::Sample sample_sample() {
    flowrt_app::Sample value{};
    std::memset(&value, 0, sizeof(value));
    value.value = std::uint32_t{2};
    return value;
}

void test_sample_message_abi() {
    static_assert(std::is_standard_layout_v<flowrt_app::Sample>);
    static_assert(std::is_trivially_copyable_v<flowrt_app::Sample>);
    static_assert(sizeof(flowrt_app::Sample) == 4);
    static_assert(alignof(flowrt_app::Sample) == 4);
    assert_default_bytes_zero<flowrt_app::Sample>();
    static_assert(offsetof(flowrt_app::Sample, value) == 0);
    assert_byte_roundtrip(sample_sample());
    assert_sample_bytes(sample_sample(), EXPECTED_SAMPLE_BYTES);
    write_fixture("sample.bin", bytes_of(sample_sample()));
}

}  // namespace

int main() {
    test_sample_message_abi();
    return 0;
}

File: crates/flowrt-codegen/tests/golden/fault_injection_restart_rust/expected/rust/tests/message_abi.rs
// FlowRT 管理产物。不要手工修改。

fn bytes_of<T>(value: &T) -> Vec<u8> {
    let mut bytes = vec![0u8; std::mem::size_of::<T>()];
    // Safety：生成测试只传入 FlowRT ABI v0.1 plain-data 消息，且 padding 已初始化。
    unsafe {
        std::ptr::copy_nonoverlapping(
            (value as *const T).cast::<u8>(),
            bytes.as_mut_ptr(),
            bytes.len(),
        );
    }
    bytes
}

fn assert_default_bytes_zero<T: Copy + Default>() {
    let value = T::default();
    assert_eq!(bytes_of(&value), vec![0u8; std::mem::size_of::<T>()]);
}

fn assert_byte_roundtrip<T: Copy + Default>(value: T) {
    let bytes = bytes_of(&value);
    let mut roundtrip = T::default();
    // Safety：`roundtrip` 是有效 plain-data 存储，`bytes` 长度等于 `size_of::<T>()`。
    unsafe {
        std::ptr::copy_nonoverlapping(
            bytes.as_ptr(),
            (&mut roundtrip as *mut T).cast::<u8>(),
            bytes.len(),
        );
    }
    assert_eq!(bytes_of(&roundtrip), bytes);
}

fn assert_sample_bytes<T: Copy>(value: T, expected: &[u8]) {
    assert_eq!(bytes_of(&value), expected);
}

const EXPECTED_SAMPLE_BYTES: &[u8] = &[2, 0, 0, 0];

fn sample_sample() -> flowrt_app::messages::Sample {
    let mut value = flowrt_app::messages::Sample::default();
    value.value = 2u32;
    value
}

#[test]
fn sample_message_abi() {
    assert_eq!(std::mem::size_of::<flowrt_app::messages::Sample>(), 4);
    assert_eq!(std::mem::align_of::<flowrt_app::messages::Sample>(), 4);
    assert_default_bytes_zero::<flowrt_app::messages::Sample>();
    assert_eq!(std::mem::offset_of!(flowrt_app::messages::Sample, value), 0);
    assert_byte_roundtrip(sample_sample());
    assert_sample_bytes(sample_sample(), EXPECTED_SAMPLE_BYTES);
}

